src/dbnode/encoding/proto/common.go

```diff
@@ -165,18 +171,30 @@ func newCustomFieldState(fieldNum int, fieldType customFieldType) customFieldSta
 
 // TODO(rartoul): Improve this function to be less naive and actually explore nested messages
 // for fields that we can use our custom compression on: https://github.com/m3db/m3/issues/1471
-func customFields(s []customFieldState, schema *desc.MessageDescriptor) []customFieldState {
+func customFields(s []customFieldState, protoFields []int32, schema *desc.MessageDescriptor) ([]customFieldState, []int32) {
+	fields := schema.GetFields()
 	numCustomFields := numCustomFields(schema)
+	numProtoFields := len(fields) - numCustomFields
+
 	if cap(s) >= numCustomFields {
+		for i := range s {
+			s[i] = customFieldState{}
+		}
 		s = s[:0]
 	} else {
 		s = make([]customFieldState, 0, numCustomFields)
 	}
 
-	fields := schema.GetFields()
+	if cap(protoFields) >= numProtoFields {
+		protoFields = protoFields[:0]
+	} else {
+		protoFields = make([]int32, 0, numProtoFields)
+	}
+
 	for _, field := range fields {
 		customFieldType, ok := isCustomField(field.GetType(), field.IsRepeated())
 		if !ok {
+			protoFields = append(protoFields, field.GetNumber())
 			continue
 		}
 
```

Comment:
unrelated nit: it looks like you should be able to set `fieldState.intEncAndIter.unsighed` inside `newCustomFieldState` unless there's a penalty for doing so, since you have all of the state required in that function (and it seems a little weird to change the state immediately after creating it).

Reply:
good call